mjpc/planners/cem/planner.cc

```diff
@@ -0,0 +1,527 @@
+  trajectory_order.reserve(num_trajectory_);
+  for (int i = 0; i < num_trajectory_; i++) {
+    trajectory_order.push_back(i);
+  }
+
+  // noise gradient
+  noise_gradient.resize(num_max_parameter);
+}
+
+// reset memory to zeros
+void CEMPlanner::Reset(int horizon) {
+  // state
+  std::fill(state.begin(), state.end(), 0.0);
+  std::fill(mocap.begin(), mocap.end(), 0.0);
+  std::fill(userdata.begin(), userdata.end(), 0.0);
+  time = 0.0;
+
+  // policy parameters
+  policy.Reset(horizon);
+  previous_policy.Reset(horizon);
+
+  // scratch
+  std::fill(parameters_scratch.begin(), parameters_scratch.end(), 0.0);
+  std::fill(times_scratch.begin(), times_scratch.end(), 0.0);
+
+  // noise
+  std::fill(noise.begin(), noise.end(), 0.0);
+
+  // trajectory samples
+  for (int i = 0; i < kMaxTrajectory; i++) {
+    trajectory[i].Reset(kMaxTrajectoryHorizon);
+    candidate_policy[i].Reset(horizon);
+  }
+
+  for (const auto& d : data_) {
+    mju_zero(d->ctrl, model->nu);
+  }
+
+  // noise gradient
+  std::fill(noise_gradient.begin(), noise_gradient.end(), 0.0);
+
+  // improvement
+  improvement = 0.0;
+
+  // winner
+  winner = 0;
+}
+
+// set state
+void CEMPlanner::SetState(const State& state) {
+  state.CopyTo(this->state.data(), this->mocap.data(), this->userdata.data(),
+               &this->time);
+}
+
+int CEMPlanner::OptimizePolicyCandidates(int ncandidates, int horizon,
+                                              ThreadPool& pool) {
+  // if num_trajectory_ has changed, use it in this new iteration.
+  // num_trajectory_ might change while this function runs. Keep it constant
+  // for the duration of this function.
+  int num_trajectory = num_trajectory_;
+  ncandidates = std::min(ncandidates, num_trajectory);
+  ResizeMjData(model, pool.NumThreads());
+
+  // ----- rollout noisy policies ----- //
+  // start timer
+  auto rollouts_start = std::chrono::steady_clock::now();
+
+  // simulate noisy policies
+  this->Rollouts(num_trajectory, horizon, pool);
+
+  // sort candidate policies and trajectories by score
+  trajectory_order.clear();
+  trajectory_order.reserve(num_trajectory);
+  for (int i = 0; i < num_trajectory; i++) {
+    trajectory_order.push_back(i);
+  }
+
+  // sort so that the first ncandidates elements are the best candidates, and
+  // the rest are in an unspecified order
+  std::partial_sort(
+      trajectory_order.begin(), trajectory_order.begin() + ncandidates,
+      trajectory_order.end(), [trajectory = trajectory](int a, int b) {
+        return trajectory[a].total_return < trajectory[b].total_return;
+      });
+
+  // stop timer
+  rollouts_compute_time = GetDuration(rollouts_start);
+
+  return ncandidates;
+}
+
+// optimize nominal policy using random sampling
+void CEMPlanner::OptimizePolicy(int horizon, ThreadPool& pool) {
+  // resample nominal policy to current time
+  this->UpdateNominalPolicy(horizon);
+
+  OptimizePolicyCandidates(n_elites, horizon, pool);
```

Comment:
note: the order of these is important. it's this way in the original predictive sampling impl, but because the ordering of the trajectory candidates is done in `OptimizePolicyCandidates` and the nominal policy is updated beforehand, we need some initialization of the order. Maybe I should change this part of the API to flow more naturally?

Reply:
we can initialize trajectory_order in the constructor

Reply:
I think this is resolved by this block in `Allocate`, but let me know if it's not:
```
trajectory_order.reserve(num_trajectory_);
for (int i = 0; i < num_trajectory_; i++) {
  trajectory_order.push_back(i);
}
```